File: java/secret-handshake/src/main/java/HandshakeCalculator.java
import java.util.List;
import java.util.Collections;
import java.util.ArrayList;

class HandshakeCalculator {

    List<Signal> calculateHandshake(int number) {
        List<Signal> r = new ArrayList<>();
        if((number & 1) !=0) r.add(Signal.WINK);
        if((number & 2) !=0) r.add(Signal.DOUBLE_BLINK);
        if((number & 4) !=0) r.add(Signal.CLOSE_YOUR_EYES);
        if((number & 8) !=0) r.add(Signal.JUMP);
        if((number & 16) !=0) Collections.reverse(r);
        return r;
        
    }

}

File: kotlin/secret-handshake/src/main/kotlin/HandshakeCalculator.kt


object HandshakeCalculator {
    fun calculateHandshake(n: Int): List<Signal> = listOf(
        Pair(n and 1 != 0, Signal.WINK),
        Pair(n and 2 != 0, Signal.DOUBLE_BLINK),
        Pair(n and 4 != 0, Signal.CLOSE_YOUR_EYES),
        Pair(n and 8 != 0, Signal.JUMP)
    ).filter {
        it.first == true
    }.map {
        it.second
    }.let {
        if (n and 16 == 0) it else it.asReversed()
    }
}